### PR TITLE
feat(issue_alerts): Cache calls to snuba for `BaseEventFrequencyCondition` subclasses.

### PR DESCRIPTION
**Pull request by wedamija**
This builds on work started in https://github.com/getsentry/sentry/pull/24231. We add in the caching
to `EventFrequencyCondition` and `EventUniqueUserFrequencyCondition`. At the moment, the tests aren't
going to be 100% accurate, because snuba also caches duplicate queries. They pass though, and I've
tested this against my local snuba where I disabled the caching.

### CODE REVIEW

**Review by wedamija — COMMENTED** [review]
I added `use_cache` to the other backends for consistency, but they won't perform any caching.

**Comment by wedamija** [comment]
> I'll follow up with another pr to disable snuba caching in sentry. Or at least, give us the option
> to disable it within a test.

Actually I do need to do this in here, or merge a pr beforehand at least.

File: src/sentry/testutils/cases.py [hunk 1]
```diff
@@ -670,6 +670,20 @@ def setUp(self):
     def initialize(self, reset_snuba, call_snuba):
         self.call_snuba = call_snuba
 
+    @contextmanager
+    def disable_snuba_query_cache(self):
+        self.snuba_update_config({"use_readthrough_query_cache": 0, "use_cache": 0})
+        yield
+        self.snuba_update_config({"use_readthrough_query_cache": None, "use_cache": None})
```
Comment (wedamija) [inline src/sentry/testutils/cases.py:677]: Could potentially store the original values, not sure it matters for tests

**Review by evanh — APPROVED** [review]
Looks good to me.

**Pull request merged**